File: letni25-26/aisd/pracownie/pracownia6/F.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <stack>
#include <cstdio>
#include <tuple>
#include <random>
#include <cmath>
#include <set>

using namespace std;


tuple<int,vector<int>,int,vector<int>> hash_nums(int n, vector<int>& nums) {
    random_device rd;              
    mt19937 gen(rd());             
    uniform_int_distribution<> distA(1, 100);
    uniform_int_distribution<> dista(10, 10000);

    int Q = 9999991;

    int m = floor(n / log2(n));
    int A = distA(gen);
    vector<int> as_list(m, 1);

    unordered_map<int, int> inv;

    bool flag = true;

    vector<vector<int>> buckets;

    // divide numbers into n/logn buckets of rize roughly n/m = logn
    while (flag) {
        for (auto l : nums) {
            int g = ((A * l) % Q % m);
            inv[l] = g;
        }

        buckets = vector<vector<int>>(m);

        for (auto l : nums) {
            int g = inv[l];
            buckets[g].push_back(l);
        }

        flag = false;
        for (int g = 0; g < m; g++) {
            if ((int)buckets[g].size() > 2 * log2(n)) {
                flag = true;
                A = distA(gen);
            }
        }
    }


    vector<int> ys(m);
    for (int i = 0; i < m; i++) {
        ys[i] = 3 * buckets[i].size();
    }

    vector<vector<int>> final_ys(m);

    // cout << "buckets\n";
    // for (auto& b : buckets) {
    //     for (auto v : b) cout << v << " ";
    //     cout << "\n";
    // }

    for (int g = 0; g < m; g++) {
        auto& b = buckets[g];
        int fails = 0;
        // cout << "bucket: " << g << " ";
        // for (auto v : b) cout << v << " ";
        // cout << ys[g] << "\n";
        bool flag_inner = true;
        while (flag_inner) {
            set<int> curr_bucket;
            flag_inner = false;
            for (auto v : b) {
                int p = (as_list[g] * v) % Q % ys[g];
                if (curr_bucket.count(p)) {
                    as_list[g] = dista(gen);
                    flag_inner = true;
                    fails += 1;
                    break;
                }
                curr_bucket.insert(p);
            }
        }

        // vector<int> final_y(ys[g], 0);
        // for (auto v : b) {
        //     final_y[(as_list[g] * v) % Q % ys[g]] = v;
        // }

        // final_ys[g] = final_y;
    }

    return {m, ys, A, as_list};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    

    cin >> N;

    vector<int> nums(N);

    for(int i = 0; i < N; i ++) {
        int a;
        cin >> a;
        nums[i] = a;
    }

    auto [m,ys,A,a_list] = hash_nums(N,nums);
    
    cout << m << '\n';

    for(int i = 0; i < m; i++) {
        cout << ys[i] << ' ';
    }

    cout << '\n' << A << '\n';

    for(int i = 0; i < m; i++) {
        cout << a_list[i] << ' ';
    }

    cout << endl;
}

File: letni25-26/aisd/pracownie/pracownia6/F.py
from math import floor,sqrt,log2
from random import randint

Q = 9_999_991

def find_params(n,nums):
    m = floor(n/log2(n))#floor(sqrt(n))
    A = randint(1,5*10**7)
    as_list = [1] * m
    
    inv = dict()

    flag = True
    while flag:
        for l in nums:
            g = ((A*l) % Q % m)
            inv[l] = g

        buckets = [[] for _ in range(m)]

        for l in nums:
            g = inv[l]
            buckets[g].append(l)

        flag = False
        for g in range(m):
            if len(buckets[g]) > 2 * log2(n):
                flag = True
                A = randint(1,5*10**7)




    ys = [3*len(b) for b in buckets]

    final_ys = [0]*m
    #get them info buckets
    # print(buckets)
    for g,b in enumerate(buckets):
        fails = 0
        # print('bucket: ',g,b, ys[g])
        
        while True: #if less then we have repreats
            curr_bucket = set()
            # print(fails)
            flag = True
            for v in b:
                p = (as_list[g] * v) % Q % ys[g]
                if p in curr_bucket:
                    # print(len(curr_bucket),p)
                    as_list[g] = randint(10,10**4)
                    flag = False
                    fails += 1
                    
                    break
                curr_bucket.add(p)

            if flag:
                break

        final_y = [0]*ys[g]
        for v in b:
            final_y[(as_list[g] * v) % Q % ys[g]] = v
        
        final_ys[g] = final_y
        # if buckets[g] >= 2:
        #     as_list[g] = randint(10,10**4)
            #jakis reset
    # print('final ys', final_ys)


    return m,ys,A,as_list

import random

def generate_task():
    N = random.randint(10**4,2*10**4)
    
    ls = [random.randint(1,5*10**6) for _ in range(N)]
    
    ls = list(set(ls))

    with open('input_rand.txt', 'w') as out:
        out_string = str(N) + '\n' + ' '.join(map(str,ls))
        out.write(out_string)

def solve():

    with open('input_rand.txt','r') as inp:
        data=inp.read().splitlines()
        N = int(data[0])
        nums = [int(i) for i in data[1].split()]

        # print(N,nums)
        # mozna tez zrobić proces wybierania m dwuktornie i poprawiania A

        m,ys,A,as_list = find_params(N,nums)

        print('sum check',sum(ys), 3*N, sum(ys) <= 3*N)

        #final check
        existing = set()
        for l in nums:
            q = ((A*l) % Q % m)
            p = (as_list[q] * l) % Q % ys[q]
            if (q,p) in existing:
                print('blad')
                break
            else:
                existing.add((q,p))

        # print(m)
        # print(' '.join(map(str,ys)))
        # print(A)
        # print(' '.join(map(str,as_list)))

# for i in range(100):
#     print(f'proba {i}')
#     generate_task()
solve()

